Adds terminating running state in test manager

jeppe-style opened the pull request:
as per title

fixes https://github.com/benchflow/benchflow/issues/499

jeppe-style committed:
Adds terminating running state

VincenzoFerme commented:
@jeppe-style 

- you need to test that you go into the `TERMINATING` state. I do not see it. I need to go over a more careful review of this tomorrow.

- in #499 you write 

>waits for EM to notify that last experiment has reached state TERMINATED

How do you make sure that you wait for the **last** experiment?

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/scheduler/TestTaskScheduler.java:
```diff
@@ -330,10 +332,11 @@ private void handleRunningTestState(String testID) throws BenchFlowTestIDDoesNot
       // Exit while waiting for the Experiment Manager to notify about the scheduled
-      // experiment to be executed.
+      // experiment to be executed or aborted (terminating state)
       else if (prevTestRunningState != null && testRunningState != null
```
Are you sure this works? The point is that I check for the sequence of state in the current code, is you change taking into account this? Add some parenthesis as well.

jeppe-style replied:
maybe we call about this?

jeppe-style replied:
remove also `!= null` checks

jeppe-style replied:
and also add a check that `RunninStatesHandler.terminating(...)` was called, for instance here https://github.com/benchflow/benchflow/blob/a19d162c400ca964b60158f6144b6baa901e2003/benchflow-test-manager/application/src/test/java/cloud/benchflow/testmanager/scheduler/TestTaskSchedulerIT.java#L378

jeppe-style commented:
- calling the `TERMINATING` state is basically happening [here](https://github.com/benchflow/benchflow/pull/532/files#diff-011215c1be2fbece1ba5058f98b5577cR437) because when a test is [terminated](https://github.com/benchflow/benchflow/pull/532/files#diff-4ee5285603cbdefec7bf569d0b6b6fb9R509) in the running state the next state is `TERMINATING`
- **last** experiment should maybe be reformulated to any **running** experiment has reached the state `TERMINATED`

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/scheduler/TestTaskScheduler.java:
```diff
@@ -480,14 +495,19 @@ public void terminateTest(String testID) {
 
           }
 
+          // set to TERMINATING and wait for experiment to terminate
```
Probably this should happen before calling the `AbortRunningTask` because otherwise you could get the result from the `experiment-manager` before the switch to the `TERMINATING` state.

jeppe-style replied:
good point

jeppe-style commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/scheduler/TestTaskScheduler.java:
```diff
@@ -189,7 +190,8 @@ public synchronized void handleRunningTest(String testID) {
         // experiment to be executed. This exits before the execution of HANDLE_EXPERIMENT_RESULT
-        else if (testRunningState != null && testRunningState.equals(HANDLE_EXPERIMENT_RESULT)) {
+        else if (testRunningState != null
+            && testRunningState.equals(TestRunningState.HANDLE_EXPERIMENT_RESULT)) {
```
I think also here we need to check for `|| testRunningState == TestRunningState.TERMINATING`

jeppe-style committed:
Improves Terminating state handling
Adds test to verify RunningStatesHandler.terminating was called when expected. Cleans up TestTaskScheduler if statements. Fixes the service name

jeppe-style commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/scheduler/TestTaskScheduler.java:
```diff
@@ -480,14 +490,19 @@ public void terminateTest(String testID) {
           }
 
+          // run the TERMINATING state to terminate
+          handleRunningTest(testID);
```
this should only be called in case we are not expecting a call from the EM

VincenzoFerme replied:
And possibly from another iteration of the outer `while` loop, and not directly here

jeppe-style commented:
we should also check that `handleTerminatedState` is called when a test is terminated

jeppe-style committed:
Adds check that `handleTerminatedState` always is called when a test should be terminated and fixes related issues

VincenzoFerme commented:
@jeppe-style ready of in-progress?

jeppe-style committed:
Adds verification that all smoke tests when through the terminated state handler

jeppe-style committed:
Fixes code style issues

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/scheduler/TestTaskScheduler.java:
```diff
@@ -181,17 +191,19 @@ public synchronized void handleRunningTest(String testID) {
-            && testRunningState.equals(TestRunningState.HANDLE_EXPERIMENT_RESULT)) {
+        else if (testRunningState == TestRunningState.HANDLE_EXPERIMENT_RESULT
+            || testRunningState == TestRunningState.TERMINATING
+                && (prevTestRunningState == TestRunningState.HANDLE_EXPERIMENT_RESULT
```
I do not get why you need to check this

jeppe-style replied:
It is because in the case of `HANDLE_EXPERIMENT_RESULT` we should not execute the `TERMINATING` handler but wait for the trial to finish on Faban

VincenzoFerme replied:
We should talk about this, it is faster

jeppe-style replied:
add a comment

jeppe-style committed:
Adds comments to explain the exiting of the while loops in the TestTaskScheduler
